toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go

```diff
@@ -228,77 +232,192 @@ func findPartitions(buildDir string, diskDevice string) ([]string, []*safechroot
 		return nil, nil, err
 	}
 
-	// Look for the boot partition (i.e. EFI system partition).
-	var efiSystemPartition *diskutils.PartitionInfo
-	for _, diskPartition := range diskPartitions {
-		if diskPartition.PartitionTypeUuid == diskutils.EfiSystemPartitionUuid {
-			efiSystemPartition = &diskPartition
-			break
+	systemBootPartition, err := findSystemBootPartition(diskPartitions)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var bootLoaderPartition *diskutils.PartitionInfo
+
+	switch systemBootPartition.PartitionTypeUuid {
+	case diskutils.EfiSystemPartitionTypeUuid:
+		bootLoaderPartition, err = findBootLoaderPartitionFromEsp(systemBootPartition, diskPartitions, buildDir)
+		if err != nil {
+			return nil, nil, err
+		}
+
+	case diskutils.BiosBootPartitionTypeUuid:
+		bootLoaderPartition, err = findBootLoaderPartitionFromBiosBootPartition(systemBootPartition, diskPartitions, buildDir)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// Note: This code assumes that the bootloader is installed on the rootfs partition.
+	// However, technically the bootloader can sit on a dedicated partition separate from both the ESP
+	// and the rootfs partition.
+	mountPoints, err := findMountsFromRootfs(bootLoaderPartition, diskPartitions, buildDir)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return nil, mountPoints, nil
+}
+
+func findSystemBootPartition(diskPartitions []diskutils.PartitionInfo) (*diskutils.PartitionInfo, error) {
+	// Look for all system boot partitions, including both EFI System Paritions (ESP) and BIOS boot partitions.
+	var bootPartitions []*diskutils.PartitionInfo
+	for i := range diskPartitions {
+		diskPartition := diskPartitions[i]
+
+		switch diskPartition.PartitionTypeUuid {
+		case diskutils.EfiSystemPartitionTypeUuid, diskutils.BiosBootPartitionTypeUuid:
+			bootPartitions = append(bootPartitions, &diskPartition)
 		}
 	}
 
-	if efiSystemPartition == nil {
-		return nil, nil, fmt.Errorf("failed to find EFI system partition (%s)", diskDevice)
+	if len(bootPartitions) > 1 {
```

Comment:
Should we pick efi as a default? & maybe add an optional parameter to override bootType?

Reply:
I haven't tried an image that supports both EFI and legacy boot, yet. So, I don't know what the correct behavior is. So, I am reporting an error until if and when we decide to support such images.

Reply:
Fair